CombinatorialGames/Game/Concrete.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import CombinatorialGames.Game.Impartial
+
+/-!
+# Combinatorial games from a type of states
+
+A "concrete" game is a type of states endowed with well-founded subsequence relations for the
+left and right players. This is often a more convenient representation for a game, which can then be
+used to define a `PGame` or any of the other combinatorial game types.
+-/
+
+open PGame
+
+variable {α : Type*}
+
+/-- A "concrete" game is a type of states endowed with well-founded subsequence relations for the
+left and right players. -/
+class ConcreteGame (α : Type*) where
+  /-- The subsequence relation for the left player. -/
+  subsequentL : α → α → Prop
+  /-- The subsequence relation for the right player. -/
+  subsequentR : α → α → Prop
+  /-- The subsequence relation is well-founded. -/
+  isWellFounded_subsequent : IsWellFounded α fun a b ↦ subsequentL a b ∨ subsequentR a b
+
+namespace ConcreteGame
+variable [ConcreteGame α]
+
+scoped infix:50 " ≺ₗ " => subsequentL
+scoped infix:50 " ≺ᵣ " => subsequentR
+attribute [instance] isWellFounded_subsequent
+
+theorem subrelation_subsequentL :
+    Subrelation subsequentL fun a b : α ↦ subsequentL a b ∨ subsequentR a b :=
+  Or.inl
+theorem subrelation_subsequentR :
+    Subrelation subsequentR fun a b : α ↦ subsequentL a b ∨ subsequentR a b :=
+  Or.inr
+
+instance [ConcreteGame α] : IsWellFounded α subsequentL :=
+  subrelation_subsequentL.isWellFounded
+instance [ConcreteGame α] : IsWellFounded α subsequentR :=
+  subrelation_subsequentR.isWellFounded
+
+/-- Defines a concrete game from a single relation, to be used for both players. -/
+def ofImpartial (r : α → α → Prop) [h : IsWellFounded α r] : ConcreteGame α where
+  subsequentL := r
+  subsequentR := r
+  isWellFounded_subsequent := by convert h; rw [or_self]
+
+/-- Turns a state of a `ConcreteGame` into a `PGame`. -/
+def toPGame (a : α) : PGame :=
+  PGame.mk {b // b ≺ₗ a} {b // b ≺ᵣ a}
+    (fun b ↦ have := subrelation_subsequentL b.2; toPGame b)
+    (fun b ↦ have := subrelation_subsequentR b.2; toPGame b)
+termination_by isWellFounded_subsequent.wf.wrap a
+
+/-- Use `toLeftMovesPGame` to cast between these two types. -/
+theorem leftMoves_toPGame (a : α) : (toPGame a).LeftMoves = {b // b ≺ₗ a} := by
+  rw [toPGame]; rfl
+/-- Use `toRightMovesPGame` to cast between these two types. -/
+theorem rightMoves_toPGame (a : α) : (toPGame a).RightMoves = {b // b ≺ᵣ a} := by
+  rw [toPGame]; rfl
+
+theorem moveLeft_toPGame_hEq (a : α) :
+    HEq (toPGame a).moveLeft fun i : {b // b ≺ₗ a} => toPGame i.1 := by
```

Comment:
Good use of HEq here!

Reply:
I copied this hack over from the nim code; I've got no idea how I came up with it back then.